### PR TITLE
feat: base token integration tests

### PR DESCRIPTION
**Pull request by cytadela8**
## What ❔

- Introduce 2 new tests for base token:
    - gas calculation
    - proper application of dynamic changes to base token ratio
- Improve forced price client fluctuations implementation to limit difference between consecutive values.

## Why ❔

- We want better integration test coverage of custom base token

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.

### CODE REVIEW

File: core/node/base_token_adjuster/src/base_token_ratio_persister.rs
Comment (cytadela8) [inline core/node/base_token_adjuster/src/base_token_ratio_persister.rs:39]: This shouldn't be needed `The first tick completes immediately.`

File: core/node/external_proof_integration_api/src/lib.rs
Comment (cytadela8) [inline core/node/external_proof_integration_api/src/lib.rs:8]: This was done by `zk fmt` :shrug:

File: core/tests/ts-integration/tests/fees.test.ts
Comment (cytadela8) [inline core/tests/ts-integration/tests/fees.test.ts:104]: This may not be needed now.

File: core/lib/external_price_api/src/forced_price_client.rs
Comment (ischasny) [inline core/lib/external_price_api/src/forced_price_client.rs:14]: Lets keep it as a configurable parameter please. Different testnets might want their variations to be different.

File: core/lib/external_price_api/src/forced_price_client.rs
Comment (ischasny) [inline core/lib/external_price_api/src/forced_price_client.rs:60]: Adjusting only numerator might not work here. Imagine USDC as a CBT. If one sets up its ratio to 1:2500 - then it'd never be adjusted. Would be great to either adjust both numerator / denominator as it was before - or at least to check which one is adjustable before choosing which one to adjust.

File: core/tests/ts-integration/tests/fees.test.ts
Comment (ischasny) [inline core/tests/ts-integration/tests/fees.test.ts:252]: typo "Denominator"

Comment (cytadela8) [reply core/tests/ts-integration/tests/fees.test.ts:252]: fixed

File: core/lib/external_price_api/src/forced_price_client.rs
Comment (cytadela8) [reply core/lib/external_price_api/src/forced_price_client.rs:60]: I would rather display a warning or silently multiply both denominator and numerator by something like `100`.  Adjusting the denominator by X% has a weird impact on the ratio, because it changes the ratio by factor of `1/(1+X)`. Enforcing next value variation limit becomes harder.

File: core/tests/ts-integration/tests/fees.test.ts
Comment (cytadela8) [reply core/tests/ts-integration/tests/fees.test.ts:264]: Lowered sleep below to 10*0.5

File: core/tests/ts-integration/tests/fees.test.ts
Comment (cytadela8) [reply core/tests/ts-integration/tests/fees.test.ts:102]: Turns out it is.

File: core/lib/external_price_api/src/forced_price_client.rs
Comment (cytadela8) [reply core/lib/external_price_api/src/forced_price_client.rs]: done

File: core/lib/external_price_api/src/forced_price_client.rs
Comment (ischasny) [reply core/lib/external_price_api/src/forced_price_client.rs:74]: Agreed, good point. Can you maybe convert it to float, calculate the next value and then decompose it to numerator / denominator as it's done [here](https://github.com/matter-labs/zksync-era/blob/main/core/lib/external_price_api/src/utils.rs#L7) for example

File: core/lib/config/src/configs/external_price_api_client.rs
Comment (ischasny) [inline core/lib/config/src/configs/external_price_api_client.rs:19]: Typo "smooth"

File: core/tests/ts-integration/tests/fees.test.ts
Comment (ischasny) [inline core/tests/ts-integration/tests/fees.test.ts:74]: Does it need to be behind `if (!isEthChain)`? As we want deposit to be made only if the chain is on custom base token

File: core/tests/ts-integration/tests/fees.test.ts
Comment (ischasny) [inline core/tests/ts-integration/tests/fees.test.ts:242]: Why not to use the same naming as we have in the forced api client already? I.e. 
 `externalPriceApiClientForcedFluctuation` for example.

File: core/tests/ts-integration/tests/fees.test.ts
Comment (ischasny) [inline core/tests/ts-integration/tests/fees.test.ts:265]: Can you please explain why `75` here?

**Review by ischasny — COMMENTED** [review]
Looks really good, please consider a few small comments / suggestions

File: core/lib/external_price_api/src/forced_price_client.rs
Comment (cytadela8) [reply core/lib/external_price_api/src/forced_price_client.rs:74]: I believe this is going to change denominator in unexpected ways. When you convert a fraction to floating point you loose information and precision.

File: core/tests/ts-integration/tests/fees.test.ts
Comment (cytadela8) [reply core/tests/ts-integration/tests/fees.test.ts:265]: updated comment

File: core/lib/external_price_api/src/forced_price_client.rs
Comment (ischasny) [reply core/lib/external_price_api/src/forced_price_client.rs:74]: But we would have to do such conversion anyway for any "real" quote providers such as coingecko, coinmarketcap and etc (i.e. they won't be supplying numerator / denominator). Also how much precision is critical for us in the forced client (given that we are allowing it to fluctuate on a fairly random number)

Comment (ischasny) [reply core/lib/external_price_api/src/forced_price_client.rs:74]: This is not something thatI'd block the PR on - but please consider converting the quote to float or multiplying onto some number as you suggested before merging.

Comment (cytadela8) [reply core/lib/external_price_api/src/forced_price_client.rs:74]: @ischasny I did add multiplying by 100 if numerator is under 100. I would rebut your argument by that setting ratio to 1:2500 is also something unrealistic as such a setting could not result from converting a "real" quote. Maybe we should go towards changing the forced_price_client config to accept a float for the price itself?

Comment (cytadela8) [reply core/lib/external_price_api/src/forced_price_client.rs:74]: Though I understand your point. I think it would be very unexpected for somebody to see the forced price client to set the price to eg. 4/10000 when they set it 1/2500. They could even see something like 7378697629483821/18446744073709551616 depending on the implementation of `Fraction::from(ratio_f64)` that is not actually documented. (The last numbers where taken from a very real python standard implementation)

File: core/tests/ts-integration/tests/fees.test.ts
Comment (perekopskiy) [inline core/tests/ts-integration/tests/fees.test.ts:203]: You shouldn't hardcode config values here, instead read them with `getFeeParams`

File: core/tests/ts-integration/tests/fees.test.ts
Comment (perekopskiy) [inline core/tests/ts-integration/tests/fees.test.ts:441]: This is no longer only about internal l1 gas price, let's rename to `setFeeParams`

File: core/tests/ts-integration/tests/fees.test.ts
Comment (perekopskiy) [inline core/tests/ts-integration/tests/fees.test.ts:244]: typo: p**oll**ing

File: core/tests/ts-integration/tests/fees.test.ts
Comment (perekopskiy) [inline core/tests/ts-integration/tests/fees.test.ts:229]: Probably I miss something, but what's the point of this check, i.e. what does it check that is not covered by other checks?

File: core/lib/external_price_api/src/forced_price_client.rs
Comment (cytadela8) [reply core/lib/external_price_api/src/forced_price_client.rs:69]: fixed

File: core/tests/ts-integration/src/retry-provider.ts
Comment (cytadela8) [reply core/tests/ts-integration/src/retry-provider.ts:63]: This specific log line is contributing >90% of lines to ZKSYNC_DEBUG_LOGS. I have placed it behind a separate env, so the logs are more useful.

File: core/tests/ts-integration/tests/fees.test.ts
Comment (perekopskiy) [reply core/tests/ts-integration/tests/fees.test.ts:203]: `getFeeParams` will return you `FeeModelConfigV2` and `BaseTokenConversionRatio`, so you can replace hardcoded `100_000_000`, `110_000_000`, `0.314` with calculated values

File: core/tests/ts-integration/tests/fees.test.ts
Comment (cytadela8) [reply core/tests/ts-integration/tests/fees.test.ts:229]: We have added this check to confirm the gas price changes when the ratio is changed.

Comment (perekopskiy) [reply core/tests/ts-integration/tests/fees.test.ts]: Well, the previous assert checks that gas price depends on the ratio. I wouldn't be against this check unless it required server restart and thus adds 30s of testing time. IMO it isn't worth it. @popzxc WDYT?

Comment (popzxc) [reply core/tests/ts-integration/tests/fees.test.ts]: Agreed.

File: core/tests/ts-integration/tests/fees.test.ts
Comment (cytadela8) [reply core/tests/ts-integration/tests/fees.test.ts]: done

File: core/tests/ts-integration/tests/fees.test.ts
Comment (cytadela8) [reply core/tests/ts-integration/tests/fees.test.ts]: done

**Comment by perekopskiy** [comment]
it seems `base ERC20 contract checks › Base token ratio is updated on L1` is flaky (check last run logs)

**Comment by cytadela8** [comment]
Due to removal of `zk` based tests, this PR is blocked by https://github.com/matter-labs/zksync-era/pull/2898

**Comment by perekopskiy** [comment]
@Deniallugo ^

**Comment by cytadela8** [comment]
> it seems `base ERC20 contract checks › Base token ratio is updated on L1` is flaky (check last run logs)

I believe I found where the price was not set properly. Seems this was not "flaky", but permanently broken. pushing a fix...

File: core/tests/ts-integration/src/utils.ts
Comment (Deniallugo) [inline core/tests/ts-integration/src/utils.ts:294]: that is obvious, but a little bit tricky part, because js sometimes treat hex values as numbers and store later as number. and it could ruin the parsing logic

File: core/tests/ts-integration/tests/fees.test.ts
Comment (perekopskiy) [inline core/tests/ts-integration/tests/fees.test.ts:265]: Multiplier is not guaranteed to be updated after each miniblock, it's updated async. Do I miss something? Does it make sense to poll L1 and wait for multiplier to be updated?

**Review by ischasny — DISMISSED** [review]

File: core/tests/ts-integration/tests/fees.test.ts
Comment (cytadela8) [reply core/tests/ts-integration/tests/fees.test.ts:265]: This is a bad comment, it should be `wait 1 miniblock so the l1 gas price gets updated`

Comment (cytadela8) [reply core/tests/ts-integration/tests/fees.test.ts:265]: Though probably polling is also better.

File: core/tests/ts-integration/src/utils.ts
Comment (cytadela8) [reply core/tests/ts-integration/src/utils.ts:294]: I have checked that there are no unexpected changes after parsing and stringifying again:

```
  console.log
    Config differences:
     --- /tmp/original_config.yml       2024-10-23 18:06:48.290023034 +0200
    +++ /tmp/new_config.yml     2024-10-23 18:06:48.290023034 +0200
    @@ -60,7 +60,7 @@
     operations_manager:
       delay_interval: 100
     state_keeper:
    -  transaction_slots: 8192
    +  transaction_slots: 1
       block_commit_deadline_ms: 2500
       miniblock_commit_deadline_ms: 1000
       miniblock_seal_queue_capacity: 10
    @@ -157,6 +157,8 @@
         poll_period: 5
         num_samples_for_blob_base_fee_estimate: 10
         internal_pubdata_pricing_multiplier: 1
    +    internal_enforced_l1_gas_price: 2000000000
    +    internal_enforced_pubdata_price: 2000000000
       watcher:
         confirmations_for_eth_event: 0
         eth_node_poll_interval: 300
    @@ -324,19 +326,19 @@
       max_rows_to_dispatch: 100
       max_retries: 5
     base_token_adjuster:
    -  price_polling_interval_ms: 30000
    -  price_cache_update_interval_ms: 2000
    +  price_polling_interval_ms: 1000
    +  price_cache_update_interval_ms: 500
       max_tx_gas: 80000
       default_priority_fee_per_gas: 1000000
       max_acceptable_priority_fee_in_gwei: 100000000000
    -  l1_receipt_checking_sleep_ms: 30000
    +  l1_receipt_checking_sleep_ms: 1000
       l1_receipt_checking_max_attempts: 3
       l1_tx_sending_max_attempts: 3
    -  l1_tx_sending_sleep_ms: 30000
    +  l1_tx_sending_sleep_ms: 1000
       halt_on_error: false
       price_fetching_max_attempts: 3
       price_fetching_sleep_ms: 5000
    -  l1_update_deviation_percentage: 10
    +  l1_update_deviation_percentage: 0
     basic_witness_input_producer:
       db_path: /home/ert/repos/zksync-era-apu-base-token-integration/./chains/valcus/db/main/basic_witness_input_producer
       window_size: 3
    @@ -344,9 +346,10 @@
     external_price_api_client:
       source: forced
       client_timeout_ms: 10000
    -  forced_numerator: 314
    -  forced_denominator: 1000
    +  forced_numerator: 300
    +  forced_denominator: 100
       forced_next_value_fluctuation: 3
    +  forced_fluctuation: 20
     consensus:
       server_addr: 127.0.0.1:3254
       public_addr: 127.0.0.1:3254

      at NodeSpawner.writeFileConfig (src/utils.ts:312:29)
```

File: core/tests/ts-integration/tests/fees.test.ts
Comment (cytadela8) [reply core/tests/ts-integration/tests/fees.test.ts:265]: Ok, so after drafting a wait like this:

```
 let gasPriceGood = false
        for (let i = 0; i < 40; i++) {
            await sleep(0.25);
            const l1GasPrice = (await alice._providerL2().getFeeParams()).V2.l1_gas_price
            if (l1GasPrice === l1GasPrice) {
                gasPriceGood = true;
                break;
            }
            console.log(l1GasPrice)
        }
        if (!gasPriceGood)
            throw new Error("Gas price change doesn't work");
```

I'm having the test fail due to unexpected gasPrice. 

```
    expect(received).toBe(expected) // Object.is equality

    Expected: 33912000n
    Received: 43960000n

      304 |         console.log('gas price', await alice._providerL2().getGasPrice());
      305 |         console.log(await alice._providerL2().getFeeParams());
    > 306 |         expect(receipt.gasPrice).toBe(BigInt(expectedConvertedGasPrice));
          |                                  ^
      307 |     });
      308 |
      309 |     test('Test base token ratio fluctuations', async () => {

      at Object.<anonymous> (tests/fees.test.ts:306:34)
```

This means fee params checked just after the transaction do not match what was used for gasPrice for the transaction. I'm not sure what exactly is causing it / what should we wait for, but I have seen consistently that doing one transaction and waiting for it to be confirmed before doing the tests transaction results in consistent results.

File: core/lib/config/src/configs/external_price_api_client.rs
Comment (cytadela8) [reply core/lib/config/src/configs/external_price_api_client.rs:20]: I have done in like this because we still have some compatibility with serde based config parsing (from env vars?). I have separately implemented the defaults in protobuf config. What would you want to me to do if I want this to be optional and have a default? _I think we should overhaul how configs are handled, so there is one code for parsing only... Maybe we should remove serde?_

Comment (cytadela8) [reply core/lib/config/src/configs/external_price_api_client.rs:20]: It's done like I did it in other configs eg. prover_job_monitor

File: core/tests/ts-integration/tests/fees.test.ts
Comment (cytadela8) [reply core/tests/ts-integration/tests/fees.test.ts:265]: So I want to just update the comment. Unless you have some idea what I can do better here

```suggestion
        //wait 1 miniblock so the gas price gets updated
        await doTxToReceiver().then((tx) => tx.wait(1));
```

Comment (perekopskiy) [reply core/tests/ts-integration/tests/fees.test.ts:265]: fee params are updated each batch, not miniblock. I'm ok with waiting for batch (please reuse existing function to do that) and changing comment

Comment (cytadela8) [reply core/tests/ts-integration/tests/fees.test.ts]: Got it! Fixed.

File: core/lib/config/src/configs/external_price_api_client.rs
Comment (cytadela8) [reply core/lib/config/src/configs/external_price_api_client.rs:20]: As you are vactioning, I'm closing this.

**Review by artmakh — APPROVED** [review]

**Pull request merged**
